### ISSUE TITLE
Feature: Flux-tree master application

### ISSUE THREAD

**Issue by jameshcorbett** [flux-framework/flux-sched#566]
While explaining Flux-tree to me, Dong mentioned that each leaf Flux instance will export a unique FLUX_TREE_ID environment variable. Since I want the applications running under the leaf Flux instances to occasionally communicate with each other, I asked him if one of those FLUX_TREE_IDs would be 0. He said that it would, so that the application with a FLUX_TREE_ID of 0 can behave kind of like MPI rank 0--doing IO, telling the other 'ranks' when to proceed, etc.--which is exactly what I wanted.

My original question:
> Is there any chance you can guarantee that one of the FLUX_TREE_IDs is 0? I want a single leaf application (my ‘batcher’, as we called it yesterday—I’ve also renamed it in the code base) to exhibit some special behavior, and I currently achieve that by passing application a unique ID on the command line, and ID 0 corresponds to the master application. It would be really for me to use Flux-Tree if I could just replace the ID passed through the command line with FLUX_TREE_ID.

Dong's response:
> As I remember, this was exactly the case. Every instance will get a unique FLUX_TREE_ID and one of them will be 0.

**Comment by dongahn** [comment]
> As I remember, this was exactly the case. Every instance will get a unique FLUX_TREE_ID and one of them will be 0.

@jameshcorbett: As I look into the code, I realize we are exporting a hierarchical tree ID as opposed to a flat ID at the leaf instance. For example, `tree.2.0` when the leaf instance which was spawned by the 3nd instance at the top level (and the first instance at the leaf level). 

From this (and the size for each level), one can easily compute an unique id by using the same trick that converts a multi-dimensional array into a 1-D array. (https://stackoverflow.com/questions/7367770/how-to-flatten-or-index-3d-array-in-1d-array) 

But this will not produce a consecutive set of IDs at the leaf level.

So my question is: do you actually want a consecutive set of IDs at the leaf level as well (0-based)? (My guess is yes). This is also doable, but I would rather export that as a separate environment variable like `FLUX_TREE_LEAF_ID` or `FLUX_LEAF_ID`. Would this be okay?

**Comment by jameshcorbett** [comment]
> So my question is: do you actually want a consecutive set of IDs at the leaf level as well (0-based)? (My guess is yes). 

@dongahn actually all I really want is a consistent (and ideally simple) way to pick out a master application. It sounds like I could do that by picking tree.0.0.0.0... if so, I wouldn't have any need for another environment variable.

By the way, is each leaf instance going to managing at least one node? What would happen if I specify some impossible tree structure, like 8 leaf instances when there are only 4 nodes managed by the root instance?

**Comment by dongahn** [comment]
> By the way, is each leaf instance going to managing at least one node? What would happen if I specify some impossible tree structure, like 8 leaf instances when there are only 4 nodes managed by the root instance?

As is, each leaf instance manages at least one core not node. Nevertheless, it currently has a logic such that if 8 instances split 4 cores, the only 4 instances each with 1 core will run. I need to test this though once my current refactoring is done.

**Comment by jameshcorbett** [comment]
Ah, OK. So I take it there's no issue with multiple sibling instances sharing a node? There are no shared memory problems like there are when multiple tasks of a broker run on the same node?

Also, how would the other environment variables, like `FLUX_TREE_LEAF_NNODES` and `FLUX_TREE_LEAF_NCORES_PER_NODE` work with Flux instances managing cores, not nodes?

**Comment by dongahn** [comment]
@jameshcorbett: I will start to test flux-tree this afternoon. So far, I have refactored the code quite a bit to make it more production- and PR-worthy.

Given that this will be a topic for our tutorial next week, I will install a version for you to try as soon as possible before this feature is merged to to our master. 

I will get your feedback first and populate test cases before requesting this to land. My guess is it will have to be after our ECP tutorial when flux-tree can be merged into our master. 

Stay tuned.

**Comment by dongahn** [comment]
FYI -- getting close. But it turns out I need to fix https://github.com/flux-framework/flux-sched/issues/581 first to continue w/ my testing.

**Comment by jameshcorbett** [comment]
No problem. Thanks for working on this!

**Comment by dongahn** [comment]
@jameshcorbett:

I posted a WIP PR that has flux-tree (and only that): PR #582.

I also installed that on rztopaz for you to try. So this version should be available on both RZ and CZ for TOSS3 systems:

```console
rztopaz188{dahn}79: salloc -N4 -ppdebug
salloc: Granted job allocation 3497569

rztopaz4{dahn}21: srun --pty --mpi=none --mpibind=off -N 4 -n 4 /usr/global/tools/flux/toss_3_x86_64_ib/flux-0.14.x-20190129-ftree/bin/flux start -o,-S,log-filename=out

rztopaz4{dahn}29: flux tree -N4 -c36 -T2x2 --njobs=10 -o perf.out ./Jobscript.sh
rztopaz4{dahn}30: ll
total 20
-rwx------ 1 dahn tools  262 Jan 29 15:36 Jobscript.sh*
-rw-rw-r-- 1 dahn tools   18 Jan 29 15:38 log.tree.1.1.1.out
-rw-rw-r-- 1 dahn tools   18 Jan 29 15:38 log.tree.1.1.2.out
-rw-rw-r-- 1 dahn tools   18 Jan 29 15:38 log.tree.1.1.3.out
-rw-rw-r-- 1 dahn tools   18 Jan 29 15:38 log.tree.1.2.1.out
-rw-rw-r-- 1 dahn tools   18 Jan 29 15:38 log.tree.1.2.2.out
-rw-rw-r-- 1 dahn tools   18 Jan 29 15:38 log.tree.2.1.1.out
-rw-rw-r-- 1 dahn tools   18 Jan 29 15:38 log.tree.2.1.2.out
-rw-rw-r-- 1 dahn tools   18 Jan 29 15:38 log.tree.2.1.3.out
-rw-rw-r-- 1 dahn tools   18 Jan 29 15:38 log.tree.2.2.1.out
-rw-rw-r-- 1 dahn tools   18 Jan 29 15:38 log.tree.2.2.2.out
-rw-rw-r-- 1 dahn tools 8590 Jan 29 15:38 out
-rw------- 1 dahn tools  977 Jan 29 15:38 perf.out
rztopaz4{dahn}31: cat Jobscript.sh
#!/bin/sh

echo "Starting ${FLUX_TREE_ID}.${FLUX_TREE_JOBSCRIPT_INDEX}..."
flux mini run -N 1 -n 1 /bin/hostname > log.${FLUX_TREE_ID}.${FLUX_TREE_JOBSCRIPT_INDEX}.out
flux mini run -N 1 -n 1 /bin/hostname >> log.${FLUX_TREE_ID}.${FLUX_TREE_JOBSCRIPT_INDEX}.out
rztopaz4{dahn}32: cat perf.out
TreeID                   Elapse(sec)           Begin(GTD)             End(GTD)     Match(usec)           NJobs NNodes  CPN
"tree"                       8.53091   1580341088.1374638    1580341096.668371               0              10     4   36
"tree.1"                     4.55694   1580341089.8812454   1580341094.4381816               0               5     2   36
"tree.1.1"                   1.33166   1580341091.1416824    1580341092.473341               0               3     1   36
"tree.1.2"                   1.71801   1580341091.6445782   1580341093.3625915               0               2     1   36
"tree.2"                     5.19001   1580341090.3047636    1580341095.494771               0               5     2   36
"tree.2.1"                   1.38036    1580341092.384315   1580341093.7646713               0               3     1   36
"tree.2.2"                   1.63394    1580341092.790617   1580341094.4245558               0               2     1   36

```

We cab discuss this at today meeting as well.

**Comment by dongahn** [comment]
This was done by #582
